data/transactions/verify/txn.go

```diff
@@ -448,3 +456,394 @@ func (w *worksetBuilder) next() (txnGroups [][]transactions.SignedTxn) {
+
+		failed, err := batchVerifier.VerifyWithFeedback()
+		// this error can only be crypto.ErrBatchHasFailedSigs
+
+		verifiedTxnGroups := make([][]transactions.SignedTxn, 0, len(bl.txnGroups))
+		verifiedGroupCtxs := make([]*GroupContext, 0, len(bl.groupCtxs))
+		failedSigIdx := 0
+		for txgIdx := range bl.txnGroups {
+			txGroupSigFailed := false
+			// if err == nil, then all sigs are verified, no need to check for the failed
+			for err != nil && failedSigIdx < bl.messagesForTxn[txgIdx] {
+				if failed[failedSigIdx] {
+					// if there is a failed sig check, then no need to check the rest of the
+					// sigs for this txnGroup
+					failedSigIdx = bl.messagesForTxn[txgIdx]
+					txGroupSigFailed = true
+				} else {
+					// proceed to check the next sig belonging to this txnGroup
+					failedSigIdx++
+				}
+			}
+			var result error
+			if !txGroupSigFailed {
+				verifiedTxnGroups = append(verifiedTxnGroups, bl.txnGroups[txgIdx])
+				verifiedGroupCtxs = append(verifiedGroupCtxs, bl.groupCtxs[txgIdx])
+			} else {
+				result = ErrInvalidSignature
+			}
+			sv.sendResult(bl.txnGroups[txgIdx], bl.elementContext[txgIdx], result)
+		}
+		// loading them all at once by locking the cache once
+		sv.cache.AddPayset(verifiedTxnGroups, verifiedGroupCtxs)
+		return struct{}{}
+	}
+
+	// if the task has an allocated seat, should release it
+	var retChan chan interface{}
+	if returnSeat {
+		retChan = sv.seatReturnChan
+	}
+	// EnqueueBacklog returns an error when the context is canceled
+	err := sv.verificationPool.EnqueueBacklog(sv.ctx, function, &uel, retChan)
+	return err
+}
+
+func getNumberOfBatchableSigsInGroup(stxs []transactions.SignedTxn) (batchSigs uint64, err error) {
+	batchSigs = 0
+	for _, stx := range stxs {
+		count, err := getNumberOfBatchableSigsInTxn(stx)
+		if err != nil {
+			return 0, err
+		}
+		batchSigs = batchSigs + count
+	}
+	return
+}
+
+func getNumberOfBatchableSigsInTxn(stx transactions.SignedTxn) (batchSigs uint64, err error) {
+	var hasSig, hasMsig bool
+	numSigs := 0
+	if stx.Sig != (crypto.Signature{}) {
+		numSigs++
+		hasSig = true
+	}
+	if !stx.Msig.Blank() {
+		numSigs++
+		hasMsig = true
+	}
+	if !stx.Lsig.Blank() {
+		numSigs++
+	}
+
+	if numSigs == 0 {
+		// Special case: special sender address can issue special transaction
+		// types (state proof txn) without any signature.  The well-formed
+		// check ensures that this transaction cannot pay any fee, and
+		// cannot have any other interesting fields, except for the state proof payload.
+		if stx.Txn.Sender == transactions.StateProofSender && stx.Txn.Type == protocol.StateProofTx {
+			return 0, nil
+		}
+		return 0, errors.New("signedtxn has no sig")
+	}
+	if numSigs != 1 {
+		return 0, errors.New("signedtxn should only have one of Sig or Msig or LogicSig")
+	}
+	if hasSig {
+		return 1, nil
+	}
+	if hasMsig {
+		sig := stx.Msig
+		for _, subsigi := range sig.Subsigs {
+			if (subsigi.Sig != crypto.Signature{}) {
+				batchSigs++
+			}
+		}
+	}
+	return
```

Comment:
if a Txn has an `Lsig` only, the function will pass through all your `if` clauses and hit this return, which I would guess implicitly returns `0, nil`. Could we make the return for Lsig-only more explicit?

Maybe there's a reason that can't happen?

Reply:
Pending a test for Logic sig, I don't think there is an issue here. 
The logic sig signatures are not added to the top level batch, and this is something to do in phase 2. 
So, if a txn has a logic sig, then there are 0 batchable signatures, and this will directly go to verification on its own.